Automattic/woocommerce-payments
# Refactor: decouple webhook processing service

A part of this PR #3827

From this discussion here https://github.com/Automattic/woocommerce-payments/pull/3827#discussion_r812854891, we decided to decouple the current webhook controller to a separate service. 

The new service will be used to process data in the current webhook controller as well as the new webhook reliability service in PR #3827


#### Changes proposed in this Pull Request









- Decouple the processing part of `WC_REST_Payments_Webhook_Controller` to a separate service `WC_Payments_Webhook_Processing_Service`: 
     - Move the processing part in WC_REST_Payments_Webhook_Controller::handle_webhook() to the new service. 
     - Change the REST-specific exception `Rest_Request_Exception` to another webhook-specific exception `Invalid_Webhook_Data_Exception`. 
     - Change name `read_rest_property()` to `read_webhook_property()`
     - Move the code handling event type `wcpay.notification`. to a new private method `process_wcpay_notification`: convert the exception here. 

- Create a new test class for the service above: WC_Payments_Webhook_Processing_Service_Test
     - Move all tests from WC_REST_Payments_Webhook_Controller_Test to this test.
     - Update tests to not rely on the rest response. Instead, focus on the mock and Invalid_Webhook_Data_Exception to decide if all webhook processes are run properly.

- Add new tests for WC_REST_Payments_Webhook_Controller_Test to focus on the behavior of WC_REST_Payments_Webhook_Controller::handle_webhook().

#### Testing instructions





* Mostly check tests and confirm they're all passed. 
* If possible, create some webhook events and confirm there is no error in the PHP error log. 

-------------------


**Post merge**
